albumentations/augmentations/geometric/functional.py

```diff
@@ -461,7 +461,9 @@ def keypoint_affine(
 
     x, y, a, s = keypoint[:4]
     x, y = skimage.transform.matrix_transform(np.array([[x, y]]), matrix.params).ravel()
-    a += rotation2DMatrixToEulerAngles(matrix.params[:2])
+    # Look to issue https://github.com/albumentations-team/albumentations/issues/1079
+    # OpenCV and skimage work differently with angle, so we need change the sign of angle
+    a -= rotation2DMatrixToEulerAngles(matrix.params[:2])
```

Comment:
Ideally, to fix this problem, we should only have to modify the code that creates the matrices; there should be no need to modify code that use the matrices in any way because the discrepancy is just in the definition of "+ve" for a rotation, and everything else should just work.

Thus, I wonder why this is necessary, and I am worried this change is incorrect and will cause problems for other use cases. I will do some experiments tomorrow to try to find some, or convince myself that it is in fact correct (I am going to test shearing and non-square images at 20 degrees; testing square images at 45 degrees can hide all sorts of bugs, as I've found out in my own projects)

In any case, if this change is correct, then it should be applied in the `rotation2DMatrixToEulerAngles` function so that it works for PerspectiveTransform too, right?

Reply:
I thought about it more, and mathematically speaking, `rotation2DMatrixToEulerAngles` should definitely return the negative of what it currently does. Currently it operates in mathland, but it should operate in imageland to be compatible with all other rotations. e.g. we would expect (y, x) = (-1, 1) to give 45 degrees in imageland.

```python
math.degrees(math.atan2(1, 1))
# 45.0
math.degrees(math.atan2(-1, 1))
# -45.0
```

Reply:
Oh, and the comment you've put is misleading. The discrepancy between OpenCV and skimage is the discrepancy between mathland and imageland. The discrepancy here is also because of that. But the discrepancy here is not because OpenCV and skimage disagree.

It's an "All A are B, but not all B are A" situation.